src/interfaces/earnings.spec.ts

```diff
@@ -0,0 +1,215 @@
+import { getAddress } from "@ethersproject/address";
+import BigNumber from "bignumber.js";
+
+import { Address, AssetHistoricEarnings, ChainId, EarningsInterface, SdkError, Usdc, VaultStatic } from "..";
+import { Context } from "../context";
+import { createMockAssetHistoricEarnings, createMockAssetStaticVaultV2 } from "../test-utils/factories";
+import { Yearn } from "../yearn";
+
+const getAddressMock = jest.fn();
+const subgraphFetchQueryMock = jest.fn();
+const oracleGetPriceUsdcMock: jest.Mock<Promise<Usdc>> = jest.fn();
+const lensAdaptersVaultsV2AssetsStaticMock: jest.Mock<Promise<VaultStatic[]>> = jest.fn();
+const getBlockNumberMock: jest.Mock<Promise<number>> = jest.fn();
+
+jest.mock("../yearn", () => ({
+  Yearn: jest.fn().mockImplementation(() => ({
+    services: {
+      subgraph: {
+        fetchQuery: subgraphFetchQueryMock
+      },
+      vision: {},
+      lens: {
+        adapters: { vaults: { v2: { assetsStatic: lensAdaptersVaultsV2AssetsStaticMock } } }
+      },
+      oracle: {
+        getPriceUsdc: oracleGetPriceUsdcMock
+      }
+    }
+  }))
+}));
+
+jest.mock("@ethersproject/address", () => ({
+  getAddress: jest.fn().mockImplementation(() => getAddressMock)
+}));
+
+jest.mock("../context", () => ({
+  Context: jest.fn().mockImplementation(() => ({
+    provider: {
+      read: {
+        getBlockNumber: getBlockNumberMock
+      }
+    }
+  }))
+}));
+
+describe("EarningsInterface", () => {
+  let earningsInterface: EarningsInterface<1>;
+
+  let mockedYearn: Yearn<ChainId>;
+
+  beforeEach(() => {
+    mockedYearn = new (Yearn as jest.Mock<Yearn<ChainId>>)();
+    earningsInterface = new EarningsInterface(mockedYearn, 1, new Context({}));
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe("protocolEarnings", () => {
+    describe("when there is no data", () => {
+      beforeEach(() => {
+        subgraphFetchQueryMock.mockResolvedValue(undefined);
+      });
+
+      it("should return a string representing the value of BigNumber(0)", async () => {
+        const actualProtocolEarnings = await earningsInterface.protocolEarnings();
+
+        expect(actualProtocolEarnings).toEqual("0");
+      });
+    });
+
+    describe("when there is data", () => {
+      beforeEach(() => {
+        oracleGetPriceUsdcMock.mockResolvedValueOnce("2.5");
+        subgraphFetchQueryMock.mockResolvedValue({
+          data: {
+            vaults: [
+              {
+                latestUpdate: {
+                  returnsGenerated: new BigNumber(2).multipliedBy(10 ** 18)
+                },
+                token: {
+                  decimals: 18,
+                  id: "tokenId"
+                }
+              },
+              {
+                token: {
+                  decimals: 18,
+                  id: "tokenIdWithoutLatestUpdate"
```

Comment:
I think we might be introducing quite a lot of boilerplate in these kinds of mocked return values. Might be a good idea to add the dependency to autogenerate the mocks that @xgambitox suggested? https://github.com/Typescript-TDD/ts-auto-mock

The other thing is that there's a bit of mental arithmetic overhead here - 2.5 * 2 = 5. I know it's very basic but can just be an overhead for the next person who is working on it. Might be good to use variables here for 2 and 2.5 and then calculate the result from that, I know there'll have to be some type casting but still might be worth doing. Likewise in other areas where we do this

Reply:
We could add it, I'm still in favour of some automation. We just need to make sure that the stubs/mocks/factories that come out of it have somewhat reasonable dummy data.

I also agree with the arithmetic, maybe worth to extend these tests to make sure those calculations are thoroughly tested. I've basically focused on adding a happy path test to each function, but now that they are there we can definitely make the tests more exhaustive.